### PR TITLE
Initial network abstraction library

### LINKED ISSUE #348: Implement a shared HTTP API abstraction
We currently make network requests by taking a direct dependency on rust networking libraries like hyper.  Longer-term we would like all our components to use the native networking stack of the app into which they're bundled.  For example, if our components are incorporated into Fenix and Fenix is using the Gecko networking stack, then so should we.

There are some moving parts involved here that are still works in progress.  GeckoView is exposing a high-level `fetch` API here:

  https://bugzilla.mozilla.org/show_bug.cgi?id=1405081

And android-components is working on a "networking" concept abstraction here:

  https://github.com/mozilla-mobile/android-components/issues/1012

While those are being resolved, we can take a step in the right direction by hiding our own networking calls behind an appropriately-shaped abstraction, while continuing to just use hyper et al initially.

From discussion with @ncalexan and @snorp, the appropriate shape is something akin to the `fetch` API:

  https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

So let's figure out how to hide all our networking details away behind a `fetch!()` macro that we can make extensible in the future.

### LINKED ISSUE #466: Connect shared HTTP abstraction to GeckoView's Necko
This is follow-on to #348.  That issue tracks abstracting HTTP access behind a facade; this ticket tracks making one of the facade implementations use the JVM to access GeckoView's lifecycle management and Necko API (introduced in [Bug 1405081](https://bugzilla.mozilla.org/show_bug.cgi?id=1405081)).

This is a "facade sibling" to #205.

### LINKED ISSUE #832: Use network abstraction in our crates.
Part of network work, see #348.

### LINKED ISSUE #833: Hook network abstraction into build system
Involves both megazord and non-megazord work.

Part of network work, see #348.

### PR DESCRIPTION
**Pull request by thomcc**
This does a lot... more of the work than I expected to be able to do so quickly. Much of this ended up being easier than expected, because we don't have to make it work asynchronously, and because the dep cycle isn't a problem. Also, having so much time to plan and think helped a lot.

### Broad work

These should be done before landing:

- [x] Implement HTTP abstraction (#348)
- [x] Implement reqwest-based backend for HTTP abstraction (#348, or implied by it)
- [x] Implement concept-fetch backend for HTTP abstraction (#466)
- [x] Replace existing networking code with HTTP abstraction (#832)
- [x] Figure out non-megazord builds (#833)
- [x] Integrate into megazords (#833)
- [x] Manual testing and benchmarking
    - My unscientific testing found this to be surprisingly close in terms of speed whether or not we're using GV or the rust stack. This is very surprising, and might indicate a flaw in my methodology, but for now I'm going to take it at face value and say 'yay'.
- [x] More docs and unit tests (#348, I guess)

### Small TODOs that I'll get to shortly:
- More considered error handling over FFI (e.g. there are places where we fail to send exceptions from java back).
- Use a-c 0.48.0-SNAPSHOT
    - Ok, apparently we're supposed to do this but gradle gets mad when I do.

### Things to address in follow-ups
- Automated tests when backed by geckoview (#834)
- Prevent the rust implementation from being compiled in (#897)
- Possibly: Allow "direct binding" over the FFI, which should help pref some. This requires using something other than ByteBuffer.ByValue, and returning a pointer instead (similar to https://github.com/gankro/thin-vec but we don't need support for mutation from Rust).
- Possibly: use `curl-rust` instead of `reqwest` as the Rust HTTP backend. since it compiles a ton faster, has far fewer dependencies, etc
- Possibly: use alternate representation to protobufs, since `prost` causes additional copies here, and response bodies can be big. (Alternatively, use something other than prost for this)

Fixes #348 
Fixes #466
Fixes #832
Fixes #833

### CODE REVIEW

File: components/support/ffi/src/error.rs [hunk 1]
```diff
@@ -161,6 +161,28 @@ impl ExternError {
         }
     }
 
+    /// Helper for the case where we aren't exposing this back over the FFI and
+    /// we just want to warn if an error occurred, release the allocated memory.
```
Comment (rfk) [inline components/support/ffi/src/error.rs]: comment nit: "...and then release..." or similar?

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+/**
+ * Singleton allowing management of the HTTP backend
+ * used by Rust components.
+ */
+object RustHttpConfig {
+    // Protects imp/client
+    private var lock = ReentrantReadWriteLock()
+    @Volatile
+    private var client: Lazy<Client>? = null
+    // IMPORTANT: This must not ever get GCed!!
```
Comment (rfk) [inline components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: Under what circumstances might to get GCed?  In other words, what things are we depending on the calling code not to do, if any?

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            if (JNA_LIBRARY_NAME == "viaduct") {
+                // TODO Enable logging if we aren't in a megazord.
+            } else {
+                // We're in a megazord. If the build is properly configured,
+                // then this is effectively a no-op since . However, there are
+                // a lot of variables and things cargo doesn't exactly
+                // guarantee that we're relying on, so in the future it
+                // is possible that somehow a build ships that contains the
+                // rust backend by accident. Calling this means we'll still
+                // make our requests through OpenSSL.
```
Comment (rfk) [inline components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt]: The comment suggests that we'll make our requests through OpenSSL, but the code suggests it will force us to make requests over the FFI via concept-fetch; which is correct?

File: components/viaduct/src/backend/ffi.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+///
+/// Takes and returns a ffi_support::ByteBuffer. (TODO: it would be nice if we could
+/// make this take/return pointers, so that we could use JNA direct mapping. Maybe
+/// we need some kind of ThinBuffer?)
+///
+/// This is a bit weird, since it requires us to allow code on the other side of
+/// the FFI to allocate a ByteBuffer from us, but it works.
+///
+/// The code on the other side of the FFI is responsible for freeing the ByteBuffer
+/// it's passed using `viaduct_destroy_bytebuffer`.
```
Comment (rfk) [inline components/viaduct/src/backend/ffi.rs]: Out of curiosity, why is it the caller's responsibility to free the input ByteBuffer?  Since the call is synchronous, it feels like we could let the caller "borrow" the bytebuffer and then free it ourselves when it returns.  To be clear, I ask because I expect to learn something interesting, not because I think this is doing it wrong :-)

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt (hunk 3 above)
Comment (thomcc) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt]: Whoops, typo, that should say Necko.

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt (hunk 2 above)
Comment (thomcc) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: It's more a note for future maintainers of this code. Nasal demons are abound if it gets GCed, so we should be cafeful not to allow that to happen.

File: components/viaduct/src/backend/ffi.rs (hunk 4 above)
Comment (thomcc) [reply components/viaduct/src/backend/ffi.rs]: Primarily to keep things consistent with how bytebuffer is used elsewhere (and how the documentation it has says it should be used), but this also has lower peak memory usage -- instead of having a buffer for the request and the buffer around for the response at the same time, we only have the one that is active around.

File: components/sync15/src/request.rs [hunk 5]
```diff
@@ -872,13 +872,13 @@ mod test {
...
             &self,
-            body: &[u8],
+            body: Vec<u8>,
             xius: ServerTimestamp,
             batch: Option<String>,
             commit: bool,
             queue: &PostQueue<T, O>,
         ) -> Result<PostResponse> {
-            self.borrow_mut().do_post(body, xius, batch, commit, queue)
+            self.borrow_mut().do_post(&body, xius, batch, commit, queue)
```
Comment (linacambridge) [inline components/sync15/src/request.rs]: I guess we could pass `Vec<u8>` directly, but this is test code, anyway...

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt (hunk 2 above)
Comment (linacambridge) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: How do we prevent that? Does the caller need to keep `CallbackImpl` around longer than `RustHttpConfig`, or do anything more?

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    override fun available(): Int {
+        return s.size() - pos
+    }
+
+    override fun skip(n: Long): Long {
+        val toSkip = Math.min((s.size() - pos).toLong(), Math.max(n, 0L)).toInt()
+        pos += toSkip
+        return toSkip.toLong()
+    }
+    // Oh come on, these are hardly magic...
```
Comment (linacambridge) [inline components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: 😆

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }()
+
+        internal var INSTANCE: LibViaduct = try {
+            val lib = Native.loadLibrary(JNA_LIBRARY_NAME, LibViaduct::class.java) as LibViaduct
+            if (JNA_LIBRARY_NAME == "viaduct") {
+                // TODO Enable logging if we aren't in a megazord.
+            } else {
+                // We're in a megazord. If the build is properly configured,
+                // then this is effectively a no-op since . However, there are
```
Comment (linacambridge) [inline components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/LibViaduct.kt]: Since...?

File: components/viaduct/src/error.rs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    BackendError(String),
+
+    #[fail(display = "Network error: {}", _0)]
+    NetworkError(String),
+
+    #[fail(display = "The rust-components network backend must be initialized before use!")]
+    BackendNotInitialized,
+
+    /// Note: we return this if the server returns a bad URL with
+    /// it's response. This *probably* should never happen, but who knows.
```
Comment (linacambridge) [inline components/viaduct/src/error.rs]: Micro-nit: its

File: components/viaduct/src/backend/reqwest.rs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        let val = String::from_utf8_lossy(v.as_bytes()).to_string();
+        let hname = match crate::HeaderName::new(k.as_str().to_owned()) {
+            Ok(name) => name,
+            Err(e) => {
+                // Ignore headers with invalid names, since nobody can look for them anyway.
+                log::warn!("Server sent back invalid header name: '{}'", e);
+                continue;
+            }
+        };
+        // Not using Header::new intentionally, since the error it returns is
```
Comment (linacambridge) [inline components/viaduct/src/backend/reqwest.rs]: Worth having a `new_unchecked` or `new_from_response` (and `new_for_request`) method that either creates the header as-is, or returns `Option<Header>`?

File: components/viaduct/src/headers/name.rs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+/// instead. This is because it would likely come through as
+/// a network error if we emitted it for local headers, when
+/// it's actually a bug that we'd need to fix.
+#[derive(failure::Fail, Debug, Clone, PartialEq)]
+#[fail(display = "Invalid header name: {:?}", _0)]
+pub struct InvalidHeaderName(Cow<'static, str>);
+
+impl From<&'static str> for HeaderName {
+    fn from(s: &'static str) -> HeaderName {
+        match HeaderName::new(s) {
```
Comment (linacambridge) [inline components/viaduct/src/headers/name.rs]: Nit: Can we use `.expect` here and below?

File: components/viaduct/src/headers/name.rs [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    }
+}
+
+fn validate_header(mut name: Cow<'static, str>) -> Result<HeaderName, InvalidHeaderName> {
+    if name.len() == 0 {
+        return Err(invalid_header_name(name));
+    }
+    let mut need_lower_case = false;
+    for b in name.bytes() {
+        let validity = VALID_HEADER_LUT[b as usize];
```
Comment (linacambridge) [inline components/viaduct/src/headers/name.rs]: H/2 makes this _so_ much easier by requiring lowercase header names.

File: components/viaduct/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+    /// Returns an [`UnexpectedStatus`] error if `self.is_success()` is false,
+    /// otherwise returns `Ok(self)`.
+    #[inline]
+    pub fn require_success(self) -> Result<Self, UnexpectedStatus> {
+        if self.is_success() {
+            Ok(self)
+        } else {
+            Err(UnexpectedStatus {
+                method: self.request_method,
+                // XXX We probably should try and sanitize this. Replace the user id
```
Comment (linacambridge) [inline components/viaduct/src/lib.rs]: I think it makes sense to do that in the callers, but I'm not sure if Viaduct is the right place to add that sanitization.

File: components/viaduct/src/settings.rs [hunk 13]
```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::time::Duration;
+
+/// Note: reqwest allows these only to be specified per-Client. concept-fetch
+/// allows these to be specified on each call to fetch. I think it's worth
+/// keeping a single global reqwest::Client in the reqwest backend, to simplify
+/// the way we abstract away from these.
```
Comment (linacambridge) [inline components/viaduct/src/settings.rs]: I agree. We might want something like `CustomClient` eventually, but we can cross that bridge when we get to it.

File: components/support/ffi/src/lib.rs [hunk 14]
```diff
@@ -385,7 +401,7 @@ impl ByteBuffer {
         let data = buf.as_mut_ptr();
         let len = buf.len();
         assert!(
-            len == (len as i64) as usize,
+            len as i64 >= 0 && (len as u64) <= (i64::max_value() as u64),
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/support/ffi/src/lib.rs]: The previous cast was broken. I let it land as r+ with fixes, but the way the check was implemented before was insufficient to verify correctness. Now it ensures that the number is greater than zero and less than the i64 maximum.

File: components/push/communications/src/lib.rs [hunk 15]
```diff
@@ -173,21 +161,22 @@ impl Connection for ConnectHttp {
                 senderid: Some(self.options.sender_id.clone()),
             });
         }
-        let mut request = match self.client.post(&url).json(&body).send() {
+        let url = Url::parse(&url)?;
+        let request = match Request::post(url).json(&body).send() {
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/push/communications/src/lib.rs]: Possibly, I didn't consider renaming things since I just made the changes mechanically.

File: components/fxa-client/src/http_client.rs [hunk 16]
```diff
@@ -41,24 +42,18 @@ impl FxAClient for Client {
...
+        let resp = Self::make_request(request)?;
+        if resp.status == status_codes::NOT_MODIFIED {
             return Ok(None);
         }
-        let etag = resp
-            .headers()
-            .get(header::ETAG)
-            .and_then(|v| v.to_str().ok())
-            .map(|s| s.to_owned());
+        let etag = resp.headers.get(header_names::ETAG).map(|s| s.to_owned());
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/fxa-client/src/http_client.rs]: I have no clue, @eoger might know. The semantics are as before, in either case.

File: components/viaduct/src/headers/name.rs (hunk 10 above)
Comment (thomcc) [reply components/viaduct/src/headers/name.rs]: I'm intentionally not doing that so I can get what the bad header is in the string, which can't be done in expect (at least, not without allocating in every case).

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt (hunk 2 above)
Comment (rfk) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: I suggest rewording this comment along the lines of the response above, as an explicit "important note to future maintainers: if you mess around with this code, you have to make sure this cant get GCed".

File: components/support/ffi/src/lib.rs [hunk 17]
```diff
@@ -395,6 +411,17 @@ impl ByteBuffer {
...
     }
 
+    /// Convert this `ByteBuffer` into a Vec<u8>. This is the only way
+    /// to access the data from inside the buffer.
+    #[inline]
+    pub fn into_vec(self) -> Vec<u8> {
+        if self.data.is_null() {
+            vec![]
+        } else {
+            unsafe { Vec::from_raw_parts(self.data, self.len as usize, self.len as usize) }
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/support/ffi/src/lib.rs]: Sure. This code just got moved from elsewhere, but more comments don't hurt.

File: components/sync15/src/client.rs [hunk 18]
```diff
@@ -143,31 +130,30 @@ impl Sync15StorageClient {
...
     fn exec_request(&self, req: Request, require_success: bool) -> error::Result<Response> {
-        log::trace!("request: {} {}", req.method(), req.url().path());
-        let resp = self.http_client.execute(req)?;
-        log::trace!("response: {}", resp.status());
+        log::trace!("request: {} {}", req.method, req.url.path());
+        let resp = req.send()?;
+        log::trace!("response: {}", resp.status);
 
-        if require_success && !resp.status().is_success() {
+        if require_success && !resp.is_success() {
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/sync15/src/client.rs]: I thought about it, but this is less of a change.

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt (hunk 2 above)
Comment (thomcc) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: We prevent that by only assigning to this once and never un-assigning it. I'll try and make this comment clearer.

To be clear, there's no danger in the code as I currently wrote it, but is a very large caveat to any future changes to this code -- this is just an extremely important invariant, and otherwise looks like a memory leak / dead store.

File: components/push/communications/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (thomcc) [reply components/push/communications/src/lib.rs]: I moved it to past tense (requested) since `response` is already a variable used for the response body.

File: components/push/ffi/Cargo.toml [hunk 19]
```diff
@@ -31,6 +31,10 @@ features = ["sqlcipher", "limits", "functions"]
 [dependencies.sync15]
 path = "../../sync15"
 
+[features]
+rust-http-stack = ["viaduct/rust-http-stack"]
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/push/ffi/Cargo.toml]: The megazords turn off this feature (or they attempt to -- there have been some indications in `cargo` issues that this behavior might change to make builds faster, but even if they do, we should still use the FFI version since we explicitly turn it on in LibViaduct.kt).

File: components/viaduct/src/lib.rs (hunk 12 above)
Comment (rfk) [reply components/viaduct/src/lib.rs]: A big, but fun, project might be to try to use the Rust typesystem to force such sanitization on all consumers via some sort of "taint checking" system. Not suggesting we do that with any urgency, but I do know that the alternative is a long tail of places where we accidentally log/expose sensitive data to places we didn't intend.
